framework/storm-tracking-cryptographic-materials-cache.md

```diff
@@ -0,0 +1,130 @@
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Initialization
+
+On initialization of the storm tracking CMC,
+the caller MUST provide exactly what is required by a
+[Local CMC](local-cryptographic-materials-cache.md).
+
+Initialization MUST also provide
+
+- [Grace Period](#grace-period)
+- [Grace Interval](#grace-interval)
+- [FanOut](#fanout)
+- [Inflight TTL](#inflight-ttl).
+- [sleepMilli](#sleepmilli).
+
+### Grace Period
+
+A number of seconds (at least 2, default 10).
+
+If an entry will expire within this amount of time,
+attempts will be made to refresh the cache.
+
+This should be significantly less than the TTL for any item put into the cache.
+
+### Grace Interval
+
+A number of seconds (at least 1, default 1).
+
+While within the [grace period](#grace-period),
+attempts to refresh the cache are made no more often than once per interval.
+
+### FanOut
+
+A number (at least 1, default 20).
+
+The maximum number of individual keys for which lookups can be in flight.
+
+### Inflight TTL
+
+A number (at least 1, default 20).
+
+An entry that has been in flight for this long is no longer considered in flight.
+
+### SleepMilli
+
+If the implementation must block, and no more intelligent signaling is used,
+then the implementation should sleep for this many milliseconds before
+reexamining the state of the cache.
+
+## Behaviors
+
+All behaviors MUST be exactly the same as a [Local CMC](local-cryptographic-materials-cache.md),
+even if used in a multi-threaded context, with two exceptions
+
+- GetCacheEntry might return NoSuchEntry, even though there is really an entry.
+- GetCacheEntry might block for a time before returning a result.
+
+Specifics for these two exceptions are outlined below.
+
+### In Flight
+
+Any time the storm tracking CMC returns NoSuchEntry from GetCacheEntry,
+that key is said to `in flight` until that same key is written with PutCacheEntry.
+
+For each in flight key, the storm tracking CMC MUST keep track of the most recent time
+that NoSuchEntry was returned, with accuracy to the second.
+
+### PutCacheEntry
+
+PutCacheEntry MUST ensure the key is not in flight.
+
+### GetCacheEntry
+
+If GetCacheEntry is called for a key :
+
+If that key is in the cache but expired, that key MUST be removed from the cache before further processing.
+
+If that key is in the cache AND that key is not within the [Grace Period](#grace-period),
+GetCacheEntry MUST return the cache entry.
+
+If the number of things inflight is greater than or equal to the [FanOut](#fanout)
+and the key is in the cache THEN GetCacheEntry MUST return the cache entry.
+
+If the number of things inflight is greater than or equal to the [FanOut](#fanout)
+and the key is not in the cache GetCacheEntry MUST block until some other option is available.
```

Comment:
a FanOut "slot" becomes available right?

Does the block have a time out?

Reply:
The block does not have a timeout. Since fanOut has a timeout, we should never be stuck forever.